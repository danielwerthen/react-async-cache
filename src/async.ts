import {
  BehaviorSubject,
  combineLatest,
  from,
  isObservable,
  Observable,
  of,
  ReplaySubject,
  Subject,
  Subscription,
} from 'rxjs';
import {
  catchError,
  concatMap,
  debounceTime,
  finalize,
  map,
  publishReplay,
  refCount,
  skip,
  startWith,
  switchMap,
  take,
  tap,
} from 'rxjs/operators';
import { exhaustMapWithTrailing, useObservedProp } from './utils';
import { AsyncFactory, AsyncResult, SharedAsync } from './types';
import { DependencyList, useCallback, useMemo } from 'react';
import { useAsyncBase } from './useAsyncBase';

export function observeAsync<T, ERR = unknown>(
  factories: Observable<AsyncFactory<T>>,
  refreshSubject: Subject<unknown> = new Subject()
): BehaviorSubject<AsyncResult<T, ERR>> {
  const subj: BehaviorSubject<AsyncResult<T, ERR>> = new BehaviorSubject<
    AsyncResult<T, ERR>
  >({
    pending: true,
    refresh: () => {
      refreshSubject.next(0);
      return subj.pipe(skip(1), take(1)).toPromise();
    },
  });
  const obsMap = new WeakMap<Observable<unknown>, ReplaySubject<unknown>>();
  const subMap = new WeakMap<Observable<unknown>, Subscription>();
  const resolveDependant = <T>(observable: Observable<T>): Promise<T> => {
    if (!obsMap.has(observable)) {
      const published = new ReplaySubject(1);
      obsMap.set(observable, published);
      subMap.set(
        observable,
        observable
          .pipe(
            map((item, idx) => {
              if (idx !== 0) {
                refreshSubject.next(0);
              }
              return item;
            })
          )
          .subscribe(published)
      );
    }
    return (obsMap.get(observable) as Observable<T>).pipe(take(1)).toPromise();
  };
  const unsubscribeDependent = (observable: Observable<unknown>) => {
    const obs = obsMap.get(observable);
    const sub = subMap.get(observable);
    if (obs) {
      obsMap.delete(observable);
    }
    if (sub) {
      sub.unsubscribe();
      subMap.delete(observable);
    }
  };
  let prevDependants: Observable<unknown>[] = [];
  let subscription: Subscription;
  const obs = combineLatest([
    factories,
    refreshSubject.pipe(debounceTime(0), startWith(0)),
  ]).pipe(
    exhaustMapWithTrailing(([factory], index) => {
      const dependants: Observable<unknown>[] = [];
      const promise = factory(observable => {
        const obs = resolveDependant(observable);
        dependants.push(observable);
        return obs;
      });
      return (isObservable(promise)
        ? promise
        : from(promise).pipe(
            concatMap(result => {
              if (isObservable(result)) {
                return result;
              }
              return of(result);
            })
          )
      ).pipe(
        finalize(() => {
          prevDependants.forEach(dep => {
            if (!dependants.includes(dep)) {
              unsubscribeDependent(dep);
            }
          });
          prevDependants = dependants;
        }),
        map(result => ({
          pending: false,
          result,
          refresh: subj.value.refresh,
        })),
        catchError(error =>
          of({
            pending: false,
            error: error as ERR,
            refresh: subj.value.refresh,
          })
        ),
        index > 0
          ? startWith({
              pending: true,
              result: subj.value.result,
              refresh: subj.value.refresh,
            })
          : tap(() => void 0)
      );
    }),
    finalize(() => {
      prevDependants.forEach(dep => {
        unsubscribeDependent(dep);
      });
      if (subscription) {
        subscription.unsubscribe();
        refreshSubject.complete();
      }
    })
  );
  subscription = obs.subscribe(subj);
  const oldComplete = subj.complete;
  return Object.assign(subj, {
    complete() {
      oldComplete.apply(subj);
      subscription.unsubscribe();
    },
  });
}

/**
 * `useAsync` will only recompute the async result when one of the `dependencies` has changed.
 * The factory function gets an observe function as its first parameter. This function can turn an
 * observable into a promise, and update the async result once the observable emits a new value.
 * @param factory Async factory method which will be provided with the observe function.
 * @param dependencies
 */
export function useAsync<T, ERR>(
  factory: AsyncFactory<T>,
  dependencies: DependencyList
): AsyncResult<T, ERR> {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  const subject = useMemo(() => {
    return observeAsync<T, ERR>(factories);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [factories]);
  return useAsyncBase(subject, subject.value);
}

/**
 * Compared to `useAsync`, this will allow the resulting value be shared by any components.
 * The resulting observable is published and refcounted to ensure the same async result is shared between subscribers.
 * The factory will be trigged when the observable is resubscribed after a period of no subscriptions.
 * @param factory Async factory method which will be provided with the observe function.
 * @param finalizeFn The resulting observable will call this callback whenever there is no subscriptions ongoing.
 */
export function shareAsync<OUTPUT, ERR = unknown>(
  factory: AsyncFactory<OUTPUT>,
  finalizeFn: () => void = () => void 0
): SharedAsync<OUTPUT, ERR> {
  const obs = of(0).pipe(
    switchMap(() => {
      return observeAsync<OUTPUT, ERR>(of(factory));
    }),
    skip(1),
    finalize(finalizeFn),
    publishReplay(1),
    refCount()
  );
  const refresh = () => {
    return obs
      .pipe(take(1))
      .toPromise()
      .then(item => item.refresh());
  };
  return Object.assign(obs.pipe(startWith({ pending: true, refresh })), {
    useSubscribe() {
      return useAsyncBase(obs, { pending: true, refresh });
    },
    refresh,
  });
}

/**
 * Sometimes you want to share a set of similar async observables, cached by a `string`. Otherwise same as `shareAsync`.
 * @param options
 */
export function sharedAsyncMap<INPUT, OUTPUT, ERR = unknown>(
  factory: (input: INPUT) => AsyncFactory<OUTPUT>,
  getKey: (input: INPUT) => string
): (input: INPUT) => SharedAsync<OUTPUT, ERR> {
  const store: {
    [key: string]: SharedAsync<OUTPUT, ERR>;
  } = {};
  const getResource = (input: INPUT) => {
    const key = getKey(input);
    if (!store[key]) {
      const res = shareAsync<OUTPUT, ERR>(factory(input), () => {
        delete store[key];
      });
      store[key] = res;
    }
    return store[key];
  };
  return getResource;
}
