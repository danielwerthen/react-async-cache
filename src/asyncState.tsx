import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
} from 'react';
import { BehaviorSubject, from, isObservable, Subject } from 'rxjs';
import {
  concatMap,
  distinctUntilChanged,
  finalize,
  map,
  tap,
} from 'rxjs/operators';
import {
  AsyncReducer,
  AsyncState,
  AsyncAction,
  AsyncStateContext,
} from './types';
import { useInitialize, useSubscribe } from './utils';

/**
 * `asyncState` allows you to share redux-like state without the boilerplate. It utilies a queue to ensure that any dispatched action
 * is reflect in the shared state in a sequential way. If two actions are dispatched at the same time, the first one will be exhausted before the next action is updating the state.
 * The dispatch function returns a promise resolves once the entire dispatched action has been exhausted.
 * An action can be a single state value, or a function that returns a single, promise or observable of state values.
 * @param initialValue
 * @param reducer
 */
export function observeAsyncState<STATE, ACTION = STATE>(
  initialValue: STATE,
  reducer: AsyncReducer<STATE, ACTION>
): AsyncState<STATE, ACTION> {
  const queue = new Subject<[AsyncAction<STATE, ACTION>, (r: STATE) => void]>();
  const state$ = new BehaviorSubject<STATE>(initialValue);
  const sub = queue
    .pipe(
      concatMap(([applicator, resolver]) => {
        const action =
          applicator instanceof Function
            ? applicator(state$.value)
            : applicator;
        let lastValue: STATE;
        const observableAction = isObservable(action)
          ? action
          : from(Promise.resolve(action));
        return observableAction.pipe(
          concatMap(action => {
            const next = reducer(state$.value, action);
            return isObservable(next) ? next : Promise.resolve(next);
          }),
          tap(res => (lastValue = res)),
          finalize(() => resolver(lastValue))
        );
      })
    )
    .subscribe(state$);
  return Object.assign(state$.asObservable(), {
    dispatch(action: AsyncAction<STATE, ACTION>): Promise<STATE> {
      return new Promise<STATE>(resolve => {
        queue.next([action, resolve]);
      });
    },
    getValue(): STATE {
      return state$.value;
    },
    useSubscribe(): STATE {
      return useSubscribe(state$, state$.value);
    },
    useSelect<OUTPUT>(
      selector: (state: STATE) => OUTPUT,
      deps: unknown[]
    ): OUTPUT {
      const callback = useCallback(selector, deps);
      const selected = useMemo(
        () => state$.pipe(map(callback), distinctUntilChanged()),
        [callback]
      );
      const initial = useInitialize(() => selector(state$.value));
      return useSubscribe(selected, initial);
    },
    unsubscribe() {
      sub.unsubscribe();
      state$.complete();
      queue.unsubscribe();
    },
  });
}

export const shareAsyncState = observeAsyncState;

export function useAsyncState<STATE, ACTION = STATE>(
  initialValue: STATE,
  reducer: AsyncReducer<STATE, ACTION>,
  deps: unknown[]
) {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const state$ = useMemo(() => observeAsyncState(initialValue, reducer), [
    reducer,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    ...deps,
  ]);
  const state = state$.useSubscribe();
  return [state, state$.dispatch];
}

/**
 * Sometime you want encapsulate the `asyncState` in a context, to allow for concurrent async states across an application.
 * @param stateFactory
 */
export function asyncStateContext<STATE, ACTION>(
  stateFactory: (initial?: STATE) => AsyncState<STATE, ACTION>
): AsyncStateContext<STATE, ACTION> {
  const context = createContext(stateFactory());
  const Provider: React.FC<{ initialState?: STATE }> = ({
    children,
    initialState,
  }) => {
    const ctx = useInitialize(() => stateFactory(initialState));
    useEffect(() => () => ctx.unsubscribe(), [ctx]);
    return <context.Provider value={ctx}>{children}</context.Provider>;
  };
  return {
    Provider,
    useSubscribe: () => {
      const state = useContext(context);
      return state.useSubscribe();
    },
    useSelect<OUTPUT>(
      selector: (state: STATE) => OUTPUT,
      deps: unknown[]
    ): OUTPUT {
      const state = useContext(context);
      return state.useSelect(selector, deps);
    },
    useDispatch: () => {
      const state = useContext(context);
      const dispatchRef = useRef(state.dispatch);
      dispatchRef.current = state.dispatch;
      return useCallback((...args) => dispatchRef.current(...args), [
        dispatchRef,
      ]);
    },
  };
}
