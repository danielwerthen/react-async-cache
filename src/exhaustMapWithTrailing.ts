import {
  from,
  Observable,
  ObservableInput,
  OperatorFunction,
  Subject,
} from 'rxjs';
import { exhaustMap, finalize, throttle } from 'rxjs/operators';

/**
 * Credit to: https://github.com/ReactiveX/rxjs/issues/5004
 * @param project
 */
export function exhaustMapWithTrailing<T, R>(
  project: (value: T, index: number) => ObservableInput<R>
): OperatorFunction<T, R> {
  return (source): Observable<R> => {
    const release = new Subject();

    return source.pipe(
      throttle(() => release, {
        leading: true,
        trailing: true,
      }),
      // TODO: Upgrade to TypeScript 3.6.3 when available and remove the cast
      // https://github.com/microsoft/TypeScript/issues/33131
      exhaustMap((value, index) =>
        from(project(value, index)).pipe(
          finalize(() => {
            release.next();
          })
        )
      ) as OperatorFunction<T, R>
    );
  };
}
