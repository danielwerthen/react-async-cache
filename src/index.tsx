import {
  catchError,
  concatMap,
  exhaustMap,
  filter,
  finalize,
  map,
  publishReplay,
  refCount,
  skip,
  startWith,
  switchMap,
  take,
  tap,
  throttle,
} from 'rxjs/operators';
import {
  BehaviorSubject,
  combineLatest,
  ConnectableObservable,
  from,
  Observable,
  ObservableInput,
  of,
  OperatorFunction,
  Subject,
  Subscription,
} from 'rxjs';
import { fromFetch as defaultFromFetch } from 'rxjs/fetch';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

/**
 * Credit to: https://github.com/ReactiveX/rxjs/issues/5004
 * @param project
 */
function exhaustMapWithTrailing<T, R>(
  project: (value: T, index: number) => ObservableInput<R>
): OperatorFunction<T, R> {
  return (source): Observable<R> => {
    const release = new Subject();

    return source.pipe(
      throttle(() => release, {
        leading: true,
        trailing: true,
      }),
      exhaustMap((value, index) =>
        from(project(value, index)).pipe(
          finalize(() => {
            release.next();
          })
        )
      )
    );
  };
}

export type AsyncResult<T, ERR> = {
  pending: boolean;
  result?: T;
  error?: ERR;
};

export type ObserveValue = <T>(input: Observable<T>) => Promise<T>;

export type AsyncFactory<T> = (observe: ObserveValue) => Promise<T>;

function getSymbol<A>(target: any, sym: symbol, factory?: () => A): A {
  if (!target[sym] && factory) {
    target[sym] = factory();
  }
  return target[sym];
}

function deleteSymbol(target: any, sym: symbol) {
  delete target[sym];
}

class FactoryObserver {
  _store = {};
  _tracker = Symbol('Factory observer tracking token');
  trigger = new Subject();

  observe<T>(input: Observable<T>, record: (sym: symbol) => void): Promise<T> {
    const token = getSymbol(input, this._tracker, () =>
      Symbol('Factory observer subscription token')
    ) as symbol;
    const [observable] = getSymbol(this._store, token, () => {
      let sub: Subscription;
      const observable = input.pipe(
        finalize(() => sub.unsubscribe()),
        publishReplay(1)
      ) as ConnectableObservable<T>;
      sub = observable
        .pipe(
          skip(1),
          tap(() => {
            setTimeout(() => {
              this.trigger.next();
            }, 0);
          })
        )
        .subscribe();
      return [observable, observable.connect()];
    }) as [ConnectableObservable<T>, Subscription];
    record(token);
    return observable.pipe(take(1)).toPromise();
  }

  isComplete(): boolean {
    const impleted = Object.getOwnPropertySymbols(this._store).map(token => {
      const stored = getSymbol(this._store, token) as [
        ConnectableObservable<unknown>,
        Subscription
      ];
      return stored[0]._isComplete;
    });
    return !impleted.some(r => !r);
  }

  unsubscribe(except?: symbol[]) {
    Object.getOwnPropertySymbols(this._store).forEach(token => {
      if (except && except.includes(token)) {
        return;
      }
      const stored = getSymbol(this._store, token) as [
        ConnectableObservable<unknown>,
        Subscription
      ];
      if (stored) {
        stored[1].unsubscribe();
      }
      deleteSymbol(this._store, token);
    });
  }
}

export function observeAsync<T, ERR = unknown>(): OperatorFunction<
  AsyncFactory<T>,
  AsyncResult<T, ERR>
> {
  return (source): Observable<AsyncResult<T, ERR>> => {
    const observer = new FactoryObserver();
    let isComplete = false;
    let lastResult: T | undefined = undefined;
    return combineLatest([
      source.pipe(finalize(() => (isComplete = true))),
      observer.trigger.pipe(startWith(0)),
    ]).pipe(
      exhaustMapWithTrailing(([factory]) => {
        const keys: symbol[] = [];
        const result = factory(input => {
          return observer.observe(input, key => keys.push(key));
        });
        if (lastResult === undefined) {
          return from(result).pipe(
            tap(item => {
              observer.unsubscribe(keys);
              lastResult = item;
            }),
            map(result => ({ pending: false, result })),
            catchError(err => of({ pending: false, error: err as ERR }))
          );
        }
        return from(result).pipe(
          tap(item => {
            observer.unsubscribe(keys);
            lastResult = item;
          }),
          map(result => ({ pending: false, result })),
          startWith({ pending: true, result: lastResult }),
          catchError(err =>
            of({ pending: false, error: err as ERR, result: lastResult })
          )
        );
      }),
      tap(() => {
        if (observer.isComplete() && isComplete) {
          observer.trigger.complete();
        }
      }),

      finalize<AsyncResult<T, ERR>>(() => {
        observer.unsubscribe();
        observer.trigger.complete();
      })
    );
  };
}

class Monitor {
  usingPending = false;
  usingError = false;

  wrap<T, ERR>(result: AsyncResultWithRefresh<T, ERR>) {
    if (this.usingPending && this.usingError) {
      return result;
    }
    return new Proxy(result, {
      get: (target, prop, receiver) => {
        switch (prop) {
          case 'pending':
            this.usingPending = true;
            break;
          case 'error':
            this.usingError = true;
            break;
        }
        return Reflect.get(target, prop, receiver);
      },
    });
  }
}

export type AsyncResultWithRefresh<T, ERR> = AsyncResult<T, ERR> & {
  refresh: () => Promise<T>;
};

export function observeAsyncWithRefresh<T, ERR = unknown>(
  factories: Observable<AsyncFactory<T>>,
  finalizeFn: () => void = () => void 0
): [Observable<AsyncResult<T, ERR>>, () => Promise<T>] {
  const refreshTrigger = new Subject();
  const root = refreshTrigger.pipe(
    startWith(0),
    switchMap(() => factories),
    observeAsync<T, ERR>(),
    finalize(finalizeFn),
    publishReplay(1),
    refCount()
  );
  const refresh = () => {
    function guard(item: any): item is T {
      return item !== undefined;
    }
    const promise = root
      .pipe(
        skip(1),
        map(item => item.result),
        filter<T | undefined, T>(guard),
        take(1)
      )
      .toPromise();
    refreshTrigger.next();
    return promise;
  };
  return [root, refresh];
}

export function useAsync<T, ERR>(
  factory: AsyncFactory<T>,
  dependencies: unknown[]
): AsyncResultWithRefresh<T, ERR> {
  const callback = useCallback(factory, dependencies);
  const monitor = useMemo(() => new Monitor(), []);
  const factories = useObservedProp(callback);
  const [observable, handleRefresh] = useMemo(() => {
    return observeAsyncWithRefresh<T, ERR>(factories);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [factories]);
  const [output, setOutput] = useState<AsyncResult<T, ERR>>({ pending: true });
  const outputRef = useRef<AsyncResult<T, ERR>>(output);
  outputRef.current = output;
  useEffect(() => {
    const sub = observable.subscribe(item => {
      if (item.pending) {
        if (monitor.usingPending) {
          setOutput(item);
        }
      } else {
        const old = outputRef.current;
        if (
          old.result !== item.result ||
          old.error !== item.error ||
          old.pending !== item.pending
        ) {
          setOutput(item);
        }
      }
    });
    return () => sub.unsubscribe();
  }, [observable, monitor, outputRef]);
  const final = useMemo(
    () =>
      monitor.wrap({
        ...output,
        refresh: handleRefresh,
      }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [monitor, output, handleRefresh]
  );
  if (!monitor.usingError) {
    if (output.error !== undefined) {
      throw output.error;
    }
  }
  return final;
}

export function useSubscribe<T>(observable: Observable<T>, initialValue: T): T {
  const [state, setState] = useState(initialValue);
  useEffect(() => {
    const sub = observable.subscribe(setState);
    return () => sub.unsubscribe();
  }, [observable]);
  return state;
}

export type SharedAsync<OUTPUT, ERR = unknown> = Observable<
  AsyncResult<OUTPUT, ERR>
> & {
  useSubscribe(): AsyncResult<OUTPUT, ERR>;
  refresh(): Promise<OUTPUT>;
  unsubscribe(): void;
};

export function shareAsync<OUTPUT, ERR = unknown>(
  factory: AsyncFactory<OUTPUT>,
  finalizeFn: () => void = () => void 0
): SharedAsync<OUTPUT, ERR> {
  const [observable, refresh] = observeAsyncWithRefresh<OUTPUT, ERR>(
    of(factory),
    finalizeFn
  );
  const subject = new BehaviorSubject({ pending: true });
  const sub = observable.subscribe(subject);
  return Object.assign(observable, {
    useSubscribe() {
      return useSubscribe(subject, subject.value);
    },
    refresh() {
      return refresh();
    },
    unsubscribe() {
      sub.unsubscribe();
      subject.complete();
    },
  });
}

export function sharedAsyncFactory<INPUT, OUTPUT, ERR = unknown>({
  factory,
  getKey,
}: {
  factory: (input: INPUT) => AsyncFactory<OUTPUT>;
  getKey: (input: INPUT) => string;
}): (input: INPUT) => SharedAsync<OUTPUT, ERR> {
  const store: {
    [key: string]: SharedAsync<OUTPUT, ERR>;
  } = {};
  const getResource = (input: INPUT) => {
    const key = getKey(input);
    if (!store[key]) {
      const res = shareAsync<OUTPUT, ERR>(
        factory(input),
        () => delete store[key]
      );
      store[key] = res;
    }
    return store[key];
  };
  return getResource;
}

export function useObservedProp<A>(a: A): Observable<A> {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const subject = useMemo(() => new BehaviorSubject<A>(a), []);
  useEffect(() => {
    if (subject.value !== a) {
      subject.next(a);
    }
  }, [a, subject]);
  useEffect(() => () => subject.complete(), [subject]);
  return useMemo(() => subject.asObservable(), [subject]);
}

export function createSharedState<T>(initialValue: T) {
  const subject = new BehaviorSubject(initialValue);
  const setState = (value: T | ((prev: T) => T)) => {
    if (value instanceof Function) {
      subject.next(value(subject.value));
    } else {
      subject.next(value);
    }
  };
  return Object.assign(subject.asObservable(), {
    useState() {
      return [useSubscribe(subject, subject.value), setState];
    },
    setState,
  });
}

export function createSharedFetch<T, INPUT extends unknown[], ERR = unknown>(
  init$: Observable<RequestInit>,
  request: (...input: INPUT) => string,
  selector: (response: Response) => Promise<T>,
  fromFetch: typeof defaultFromFetch = defaultFromFetch
): (...input: INPUT) => SharedAsync<T, ERR> {
  const shared = sharedAsyncFactory<string, T, ERR>({
    factory(input: string) {
      return async observe => {
        const init = await observe(init$);
        return observe(
          fromFetch(input, {
            ...init,
            selector,
          })
        );
      };
    },
    getKey(input: string) {
      return input;
    },
  });
  return (...input: INPUT) => {
    return shared(request(...input));
  };
}

function isPromise<T>(obj: any): obj is Promise<T> {
  return obj && typeof obj.then === 'function';
}

function isObservable<T>(obj: any): obj is Observable<T> {
  return obj && typeof obj.subscribe === 'function';
}

export type AsyncState<T> = Observable<T> & {
  dispatch: (action: (v: T) => T | Promise<T> | Observable<T>) => Observable<T>;
  unsubscribe: () => void;
};

export function asyncState<T>(initialValue: T): AsyncState<T> {
  const queue = new Subject<
    [
      (v: T) => T | Promise<T> | Observable<T>,
      (r: T | Promise<T> | Observable<T>) => void
    ]
  >();
  const state$ = new BehaviorSubject<T>(initialValue);
  const sub = queue
    .pipe(
      concatMap(([applicator, resolver]) => {
        const next = applicator(state$.value);
        resolver(next);
        if (isObservable(next)) {
          return next;
        } else if (isPromise(next)) {
          return from(next);
        }
        return of(next);
      })
    )
    .subscribe(state$);
  return Object.assign(state$.asObservable(), {
    dispatch(action: (v: T) => T | Promise<T> | Observable<T>) {
      return from(
        new Promise<T | Promise<T> | Observable<T>>(resolve => {
          queue.next([action, resolve]);
        })
      ).pipe(
        concatMap(res => {
          if (isObservable(res)) {
            return res;
          } else if (isPromise(res)) {
            return from(res);
          }
          return of(res);
        })
      );
    },
    unsubscribe() {
      sub.unsubscribe();
      state$.complete();
    },
  });
}
