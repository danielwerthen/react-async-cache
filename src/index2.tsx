import {
  catchError,
  exhaustMap,
  filter,
  finalize,
  map,
  publishReplay,
  skip,
  startWith,
  take,
  tap,
  throttle,
} from 'rxjs/operators';
import {
  BehaviorSubject,
  combineLatest,
  ConnectableObservable,
  from,
  Observable,
  ObservableInput,
  of,
  OperatorFunction,
  Subject,
  Subscription,
} from 'rxjs';

/**
 * Credit to: https://github.com/ReactiveX/rxjs/issues/5004
 * @param project
 */
function exhaustMapWithTrailing<T, R>(
  project: (value: T, index: number) => ObservableInput<R>
): OperatorFunction<T, R> {
  return (source): Observable<R> => {
    const release = new Subject();

    return source.pipe(
      throttle(() => release, {
        leading: true,
        trailing: true,
      }),
      exhaustMap((value, index) =>
        from(project(value, index)).pipe(
          finalize(() => {
            release.next();
          })
        )
      )
    );
  };
}

export type AsyncResult<T, ERR> = {
  pending?: boolean;
  result?: T;
  error?: ERR;
  refresh: () => Promise<T>;
};

export type ObserveValue = <T>(input: ObservableInput<T>) => Promise<T>;

export type AsyncFactory<T> = (observe: ObserveValue) => Promise<T>;

function getSymbol<A>(target: any, sym: symbol, factory?: () => A): A {
  if (!target[sym] && factory) {
    target[sym] = factory();
  }
  return target[sym];
}

function deleteSymbol(target: any, sym: symbol) {
  delete target[sym];
}

class FactoryObserver {
  _store = {};
  _tracker = Symbol('Factory observer tracking token');
  trigger = new Subject();

  observe<T>(
    input: ObservableInput<T>,
    record: (sym: symbol) => void
  ): Promise<T> {
    const token = getSymbol(input, this._tracker, () =>
      Symbol('Factory observer subscription token')
    ) as symbol;
    const [observable] = getSymbol(this._store, token, () => {
      let sub: Subscription;
      const observable = from(input).pipe(
        finalize(() => sub.unsubscribe()),
        publishReplay(1)
      );
      sub = observable
        .pipe(
          skip(1),
          tap(() => this.trigger.next())
        )
        .subscribe();
      return [observable, observable.subscribe()];
    }) as [ConnectableObservable<T>, Subscription];
    record(token);
    return observable.pipe(take(1)).toPromise();
  }

  unsubscribe(except?: symbol[]) {
    Object.getOwnPropertySymbols(this._store).forEach(token => {
      if (except && except.includes(token)) {
        return;
      }
      const stored = getSymbol(this._store, token) as [
        ConnectableObservable<unknown>,
        Subscription
      ];
      if (stored) {
        stored[1].unsubscribe();
      }
      deleteSymbol(this._store, token);
    });
  }
}

export default function observeAsync<T, ERR = unknown>(
  factories: Observable<AsyncFactory<T>>
): Observable<AsyncResult<T, ERR>> {
  const observer = new FactoryObserver();
  const lastResult = new BehaviorSubject<T | undefined>(undefined);
  const refreshTrigger = new Subject();
  const refresh = () => {
    refreshTrigger.next();
    function guard(item: any): item is T {
      return item !== undefined;
    }
    return lastResult
      .pipe(skip(1), filter<T | undefined, T>(guard), take(1))
      .toPromise();
  };
  return combineLatest([
    factories,
    observer.trigger,
    refreshTrigger.pipe(startWith()),
  ]).pipe(
    exhaustMapWithTrailing(([factory]) => {
      const keys: symbol[] = [];
      const result = factory(input => {
        return observer.observe(input, key => keys.push(key));
      });
      return from(result).pipe(
        tap(() => observer.unsubscribe(keys)),
        tap(result => lastResult.next(result)),
        map(result => ({ result })),
        catchError(err => of({ error: err as ERR })),
        startWith({ pending: true }),
        map(res => ({
          ...res,
          refresh,
        }))
      );
    }),
    finalize<AsyncResult<T, ERR>>(() => observer.unsubscribe())
  );
}
