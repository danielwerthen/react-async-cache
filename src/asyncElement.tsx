import React, {
  DependencyList,
  ReactElement,
  ReactNode,
  useCallback,
  useMemo,
} from 'react';
import { observeAsync } from './async';
import { AsyncFactory } from './types';
import { useAsyncBase } from './useAsyncBase';
import { useObservedProp } from './utils';

export function useAsyncElement<T extends ReactElement<any, any> | null, ERR>(
  factory: AsyncFactory<T>,
  dependencies: DependencyList
): ReactNode {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  const subject = useMemo(() => {
    return observeAsync<T, ERR>(factories);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [factories]);
  const Comp = useCallback(() => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const { result } = useAsyncBase(subject, subject.value);
    return result || null;
  }, [subject]);

  return <Comp />;
}
