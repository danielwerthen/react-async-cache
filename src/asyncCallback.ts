import { useCallback, useMemo } from 'react';
import { Observable, of } from 'rxjs';
import { catchError, map, startWith, switchMap, take } from 'rxjs/operators';
import { observeAsyncState } from './asyncState';
import { AsyncCallback } from './types';
import { useAsyncBase } from './useAsyncBase';
import { useObservedProp } from './utils';

export function observeAsyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factories: Observable<(...input: INPUT) => Promise<T>>
) {
  const execute: (
    ...inputs: INPUT
  ) => Promise<AsyncCallback<INPUT, T, ERR>> = async (...input: INPUT) => {
    return await state$.dispatch(() => {
      return factories.pipe(
        take(1),
        switchMap(fn => fn(...input)),
        map(result => ({ pending: false, result, execute })),
        catchError(err => of({ pending: false, error: err as ERR, execute })),
        startWith({
          pending: true,
          execute,
        })
      );
    });
  };
  const state$ = observeAsyncState<AsyncCallback<INPUT, T, ERR>>(
    {
      pending: false,
      execute,
    },
    (_state, action) => action
  );
  return Object.assign(state$, {
    execute(...input: INPUT) {
      return state$
        .pipe(
          take(1),
          switchMap(item => item.execute(...input))
        )
        .toPromise();
    },
    useSubscribe() {
      return useAsyncBase(state$, state$.getValue());
    },
    useSelect<O>(
      selector: (state: AsyncCallback<INPUT, T, ERR>) => O,
      deps: unknown[]
    ) {
      return state$.useSelect(selector, deps);
    },
  });
}

export function shareAsyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factory: (...input: INPUT) => Promise<T>
) {
  return observeAsyncCallback<INPUT, T, ERR>(of(factory));
}

/**
 * Use this instead of `useAsync` if you want to control the timing of the asynchronous operations. Operations will be executed in an enforced sequence, using an underlying `asyncState`.
 * @param factory
 * @param dependencies
 */
export function useAsyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factory: (...input: INPUT) => Promise<T>,
  dependencies: unknown[]
): AsyncCallback<INPUT, T, ERR> {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const state$ = useMemo(() => observeAsyncCallback<INPUT, T, ERR>(factories), [
    factories,
  ]);
  return useAsyncBase(state$, state$.getValue());
}
