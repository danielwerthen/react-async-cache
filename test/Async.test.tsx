import { interval, Subject } from 'rxjs';
import { map, take, toArray } from 'rxjs/operators';
import { observeAsync, AsyncFactory } from '../src/index';

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe('Thing', () => {
  it('renders without crashing', async () => {
    const interval1 = interval(200).pipe(take(3));
    const interval2 = interval(250).pipe(take(3));
    const factories = new Subject<AsyncFactory<any>>();
    const observable = observeAsync(factories);
    const result = observable
      .pipe(
        map(res => res.result),
        toArray()
      )
      .toPromise();
    factories.next(async observe => {
      const [a, b] = await Promise.all([
        observe(interval1),
        observe(interval2),
      ]);
      return { a, b, version: 1 };
    });
    await sleep(300);
    factories.next(async observe => {
      const a = await observe(interval1);
      const b = await observe(interval2);
      return { a, b, version: 2 };
    });
    await sleep(300);
    factories.next(async observe => {
      const a = await observe(interval1);
      const b = await observe(interval2);
      return { c: a + b, version: 3 };
    });

    factories.complete();
    expect(await result).toMatchSnapshot();
  });
});
