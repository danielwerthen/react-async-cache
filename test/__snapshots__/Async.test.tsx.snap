// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thing renders without crashing 1`] = `
Array [
  undefined,
  Object {
    "a": 0,
    "b": 0,
    "version": 1,
  },
  Object {
    "a": 0,
    "b": 0,
    "version": 2,
  },
  Object {
    "a": 1,
    "b": 0,
    "version": 2,
  },
  Object {
    "a": 1,
    "b": 1,
    "version": 2,
  },
  Object {
    "a": 2,
    "b": 1,
    "version": 2,
  },
  Object {
    "c": 3,
    "version": 3,
  },
  Object {
    "c": 4,
    "version": 3,
  },
]
`;
